signature/src/policy.rs

```diff
@@ -0,0 +1,460 @@
+    key_path: String,
+    // KeyData contains the trusted key(s), base64-encoded.
+    // Exactly one of KeyPath and KeyData must be specified.
+    #[serde(default, rename = "keyData")]
+    key_data: String,
+
+    // SignedIdentity specifies what image identity the signature must be claiming about the image.
+    // Defaults to "match-exact" if not specified.
+    #[serde(default, rename = "signedIdentity")]
+    signed_identity: Option<PolicyReferenceMatch>,
+}
+
+impl PolicyReqSignedBy {
+    fn signature_judge_allowed(&self, image: &mut image::Image) -> Result<()> {
+        if self.r#type != PolicyReqType::TypeSignedBy.to_string() {
+            return Err(anyhow!(ErrorInfo::ErrUnknownPrType.to_string()));
+        }
+
+        let sigs = image.get_signatures()?;
+        if sigs.is_empty() {
+            return Err(anyhow!("Can not find any signatures."));
+        }
+
+        let mut reject_reason: Vec<anyhow::Error> = Vec::new();
+
+        for sig in sigs.iter() {
+            match self.is_signature_valid(image, sig.to_vec()) {
+                // One accepted signature is enough.
+                Ok(()) => {
+                    return Ok(());
+                }
+                Err(e) => {
+                    reject_reason.push(e);
+                }
+            }
+        }
+
+        Err(anyhow!(format!(
+            "The signatures do not satisfied! Reject reason: {:?}",
+            reject_reason
+        )))
+    }
+
+    #[allow(unused_assignments)]
+    fn is_signature_valid(&self, image: &image::Image, sig: Vec<u8>) -> Result<()> {
+        // FIXME: only support "GPGKeys" type now.
+        if self.key_type != KeyType::Gpg.to_string() {
+            return Err(anyhow!(
+                "UNKNOWN KEY TYPE in policy config: only support 'GPGKeys' now."
+            ));
+        }
+
+        if !self.key_path.is_empty() && !self.key_data.is_empty() {
+            return Err(anyhow!(
+                "ERROR: both keyPath and keyData specified."
+            ));
+        }
+
+        let mut pubkey_ring = Vec::new();
+        if !self.key_data.is_empty() {
+            // base64 encode the keyring data.
+            pubkey_ring = base64::decode(&self.key_data)?;
+        } else {
+            pubkey_ring = fs::read(&self.key_path)?;
+        }
+
+        // TODO: Verify the signature with the pubkey ring.
+
+        Ok(())
+    }
+}
+
+// PolicyReferenceMatch specifies a set of image identities(image-reference) accepted in PolicyRequirement.
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+#[serde(untagged)]
+pub enum PolicyReferenceMatch {
+    // MatchRemapIdentity: "remapIdentity".
+    RemapIdentity(MatchRemapIdentity),
+    // MatchExactRepository: "exactRepository".
+    ExactRepository(MatchExactRepository),
+    // MatchExactReference: "exactReference".
+    ExactReference(MatchExactReference),
+    // MatchSimple: "matchExact", "matchRepoDigestOrExact", "matchRepository".
+    Match(MatchSimple),
+}
+
```

Comment:
Here as well, the code would probably benefit from using a trait instead:

```rust
put trait PolicyReferenceMatcher {
     fn matches_docker_reference(&self, origin: &Reference, signed_reference: &str) -> Result<()>
}
```

And have a *separate* structure for the kind of matches that you currently want to support.

Reply:
As above